### DISCUSSION TITLE
MacBook Pro and iPXE

### DISCUSSION THREAD

**Discussion by cmccluskey** [ipxe/ipxe#196] [Q&A]
Hello,

My goal is to use iPXE on an small USB stick to boot from an iSCSI target (Linux) from a MacBook Pro (2012, 'MacBookPro9,1', A1286, MD104LL/A, Boot ROM: 226.0.0, SMC Version: 2.1f175, Running OSX 10.11.6 so may be behind on some of the latest Firmware). I have tried the latest code using the bin/ipxe.iso and the EFI build from bin-x86_64-efi/ipxe.usb and bin-x86_64-efi/ipxe.efi (using the instructions from https://gist.github.com/AdrianKoshka/5b6f8b6803092d8b108cda2f8034539a). Both have various results, but not a successful one yet. 

---
The ISO boots, but I cannot interact via keyboard as it stops working. I was able to set the static interface params of the net0 interface (Vendor ID: 0x14e4, Device ID: 0x16b4 -- Broadcom tg3) into an EMBED script and they seem to be accepted, but the interface tries to DHCP anyway. At the time the DHCP is started, the link isn't up, and the sanboot fails. 

1) Is there a reason why it is proceeding to DHCP when there is a static address assigned?

2) Is there a reason why it is proceeding to DHCP when the link doesn't appear to be up?

I added a DHCP server to this segment, but since the link isn't up (presumed) at the time DHCP tries, it too fails.

When I boot into OSX/Linux the link is up and fine, so know the hardware is ok here. :)

When I add a "shell" to the embed script, I do get to the iPXE prompt. But again no keyboard, so I wasn't able to proceed with the shell to debug further. 

---
Moving on to the EFI build -- this does boot up, but fails during startup -- it seems to hang after the debug line "Adding PCI Root bus" .
![IMG_3691](https://user-images.githubusercontent.com/8421790/103468585-853abf00-4d0f-11eb-8faa-a30da6c55c93.jpg)

This fails in the same place if I use the usb image or the manual EFI packing instructions from the Gist above.

I went to the EFI since there were some posts regarding Macs needing the EFI build to fix the keyboard issues. I wasn't able to get far enough to test. 

3) Is there any way that I can tell is this notebook model needs the legacy ISO or the EFI build?

4) Any recommendation on the next steps? I made the most progress with the ISO, but if there isn't a working keyboard it will be difficult to diagnose further.

Thank you for your time/help,
-Chris

Current EMBED script:
```
#!ipxe

echo Got here
#shell
set net0/ip 192.168.69.61
set net0/netmask 255.255.255.0
#set net0/gateway 0.0.0.0 
set net0/mtu 1500
sleep 5
set keep-san 1
sanboot iscsi:192.168.69.66::::iqn.2000-01.com.blahblah:blah.blah.eee261eb4 

```

**Comment by mcb30** [comment]
> ```
> #!ipxe
> 
> echo Got here
> #shell
> set net0/ip 192.168.69.61
> set net0/netmask 255.255.255.0
> #set net0/gateway 0.0.0.0 
> set net0/mtu 1500
> sleep 5
> set keep-san 1
> sanboot iscsi:192.168.69.66::::iqn.2000-01.com.blahblah:blah.blah.eee261eb4 
> ```

Easy things first: if you are trying to use a static IP configuration then you will also need to include an `ifopen net0` command to actually open the interface.  (Opening the interface is implicit when performing DHCP.)

You shouldn't need the `set net0/mtu 1500` since that is the default MTU anyway, and you shouldn't need the `set keep-san 1` since that will have no effect in your setup.  The `sleep 5` should also be unnecessary.

It's probably worth adding the `ifstat` and `route` commands immediately before `sanboot` so that you have some visibility into the applied network configuration.

So, I would suggest adjusting your embedded script to:

```
#!ipxe

ifopen net0
set net0/ip 192.168.69.61
set net0/netmask 255.255.255.0
ifstat
route
sanboot iscsi:192.168.69.66::::iqn.2000-01.com.blahblah:blah.blah.eee261eb4 
```

> The ISO boots, but I cannot interact via keyboard as it stops working. I was able to set the static interface params of the net0 interface (Vendor ID: 0x14e4, Device ID: 0x16b4 -- Broadcom tg3) into an EMBED script and they seem to be accepted, but the interface tries to DHCP anyway. At the time the DHCP is started, the link isn't up, and the sanboot fails.
> 
>  1. Is there a reason why it is proceeding to DHCP when there is a static address assigned?
>  2. Is there a reason why it is proceeding to DHCP when the link doesn't appear to be up?

There's no way that your embedded script could cause iPXE to perform DHCP.  It is most likely that either your embedded script is not actually being used, or that iPXE has exited and something else is attempting DHCP.  Could you provide a screenshot for this scenario?

> I went to the EFI since there were some posts regarding Macs needing the EFI build to fix the keyboard issues. I wasn't able to get far enough to test.
> 
> 1. Is there any way that I can tell is this notebook model needs the legacy ISO or the EFI build?

Both should work.  (I'm surprised that legacy boot works at all on a Mac, but if it's capable of getting as far as running iPXE from `bin/ipxe.iso` then everything else should also work.)

> 2. Any recommendation on the next steps? I made the most progress with the ISO, but if there isn't a working keyboard it will be difficult to diagnose further.

For the legacy ISO, I'd start by fixing the embedded script (as per above), and providing a screenshot of the failure if it still doesn't work.

For UEFI, I'd suggest building with `DEBUG=init,driver,pci,efi_driver` to try to get more detail on where it might be locking up.

**Reply by cmccluskey** [reply]
Sorry for the delay. I wanted to check my switch setup, so I changed the switch out to one that would have a faster linking time. No difference. Put original switch back in.

When I tried to add device to the DEBUG list I got the following error --
```make: *** No rule to make target `bin/driver.dbg1.o', needed by `bin/blib.a'.  Stop.``` so I omitted it.

I built the UEFI USB image with `make bin-x86_64-efi/ipxe.usb EMBED=myscript.ipxe DEBUG=device,pci,snp,snponly,snpnet,init,netdevice,pci,efi_driver,rbg,drbg`.

This locked up in the same place with additional debug info --
![UEFI-2](https://user-images.githubusercontent.com/8421790/103979169-063fef00-5132-11eb-9634-a00196123083.jpg)

Meanwhile on the ISO side, I built with similar debug -- ```make bin/ipxe.iso EMBED=myscript.ipxe DEBUG=device,pci,snp,snponly,snpnet,init,netdevice,pci,efi_driver,rbg,drbg,driver```. It looks like the interface inits fine, and I do see the link up. But sanboot failed. I tried to build once with `ping --count 3 192.168.69.69` in place of sanboot, but it failed with "Command not found."
![ISO-2](https://user-images.githubusercontent.com/8421790/103979775-58354480-5133-11eb-82c3-a764eff29006.jpg)

Both the UEFI and the ISO build both used the same embedded script -- 
```
#!ipxe

echo Got here
ifopen net0
echo After open
set net0/ip 192.168.69.61
set net0/netmask 255.255.255.0
echo Before ifstat
ifstat
route
echo Before sanboot
sanboot iscsi:192.168.69.69::::iqn.2000-01.com.synology:RAID15.Ubuntu.907e261eb4
```

Thanks for the continued help.

**Reply by mcb30** [reply]
> When I tried to add driver to the DEBUG list I got the following error --
> `` make: *** No rule to make target `bin/driver.dbg1.o', needed by `bin/blib.a'. Stop. `` so I omitted it.

My typo, sorry.  As I think you've correctly guessed, I meant `DEBUG=device` rather than `DEBUG=driver`.

> I built the UEFI USB image with `make bin-x86_64-efi/ipxe.usb EMBED=myscript.ipxe DEBUG=device,pci,snp,snponly,snpnet,init,netdevice,pci,efi_driver,rbg,drbg`.
> 
> This locked up in the same place with additional debug info --
> ![UEFI-2](https://user-images.githubusercontent.com/8421790/103979169-063fef00-5132-11eb-9634-a00196123083.jpg)

OK, so it's apparently locking up in the middle of the entropy gathering.  Entropy is required only for HTTPS (or other cryptographic operations).  It would therefore be worth checking if everything works as expected if you don't enable HTTPS in `config/general.h` (i.e. leave it in the default configuration).

To debug the entropy issue further, you can add `efi_entropy` to the `DEBUG=...` list.

> Meanwhile on the ISO side, I built with similar debug -- `make bin/ipxe.iso EMBED=myscript.ipxe DEBUG=device,pci,snp,snponly,snpnet,init,netdevice,pci,efi_driver,rbg,drbg,driver`. It looks like the interface inits fine, and I do see the link up. But sanboot failed. I tried to build once with `ping --count 3 192.168.69.69` in place of sanboot, but it failed with "Command not found."
> ![ISO-2](https://user-images.githubusercontent.com/8421790/103979775-58354480-5133-11eb-82c3-a764eff29006.jpg)

This seems to have basically worked.  You have a final error of http://ipxe.org/1d704039 which is a SCSI-level error.  This indicates that the iSCSI connection has been made, and therefore network communication is working as expected.  You should see some evidence of the connection in the iSCSI target server logs, or by capturing traffic on the iSCSI server.

Most iSCSI targets require you to define an ACL permitting access by the initiator (iPXE in this case).  Did you define an ACL on your iSCSI target and, if so, which iSCSI initiator name did you use?

**Comment by NiKiZe** [comment]
You can find a long thread in the old forum about Mac's and that the tg3 driver stopped working after some OSX version that updated firmware and got the nic in incompatible state.

**Reply by cmccluskey** [reply]
This thread or a different one -- https://forum.ipxe.org/archive/index.php?thread-8296.html?

The thread above does sound like a similar issue. But for a different model of Macintosh with a completely different Device ID and BootROM.... but possible. :)

Anything I can do to confirm this?

**Reply by NiKiZe** [reply]
All Mac's using tg3 driver most likely has the same issue, to verify, find an identical machine that has never had the is updated.

The best would be if you could get a debug session with @mcb30 and try to get it working. But please try what he wrote first just to be sure.
